### PR TITLE
[Clang Importer] Defer module imports to end of bridging-header parse.

### PR DESCRIPTION
**Pull request by graydon**
There is a subtle incompatibility between the way bridging headers handle
module imports and the way clang's normal module-loading works (as done
by -emit-pch in bridging PCH, in particular).

When importing a submodule, Swift implicitly imports its supermodule. This
is part of Swift's treatment of modules and fine, we're not changing it
here.

But if client code imports a submodule, then tries to use a type that is
only defined in its supermodule, not the submodule, this _should_ cause
a parse error (and does in clang alone, or when generating a PCH).
Unfortunately Swift's "implicit parent import" currently happens
eagerly, so the supermodule is imported and its type is defined as soon
as the submodule is imported, which in turn suppresses the error.

This in turn means that client code thinks their code "works" and then
"breaks" when they turn on bridging PCH. What _should_ happen here is
that the (actually broken) client code should not be accepted in the first
place, neither bridging PCH nor textual bridging-header import.

This commit merely changes textual bridging-header import from eager
import to deferred parent-import, like bridging PCH does. This reconciles
the difference in behaviour between the two, at the cost of a source-compat
break.

rdar://30615193

### CODE REVIEW

**Comment by graydon** [comment]
Remaining to deal with:

- [ ] A narrow test illustrating the problem & solution
- [ ] A decision on whether and how to take the source-compat break (possibly put behind swift-4 mode?)

**Comment by graydon** [comment]
@swift-ci please test

File: lib/ClangImporter/ImporterImpl.h [hunk 1]
```diff
@@ -275,6 +275,7 @@ class LLVM_LIBRARY_VISIBILITY ClangImporter::Implementation
 
   bool IsReadingBridgingPCH;
   llvm::SmallVector<clang::serialization::SubmoduleID, 2> PCHImportedSubmodules;
+  llvm::SmallVector<const clang::Module*, 2> ImportedModules;
```
Comment (jrose-apple) [inline lib/ClangImporter/ImporterImpl.h]: Can you name this DeferredHeaderImports or something? ImportedModules makes it sound like something that stays valid and makes sense to ask for for the lifetime of the ClangImporter.

**Review by jrose-apple — COMMENTED** [review]
Seems reasonable to me.

File: lib/ClangImporter/ImporterImpl.h (hunk 1 above)
Comment (graydon) [inline lib/ClangImporter/ImporterImpl.h]: sure, will do.

**Comment by graydon** [comment]
Decision was to land this for swift 4 timeframe, not swift 3.1, but consider it just a bug fix (with release note), not a behaviour to preserve under compat flag.

**Comment by graydon** [comment]
@swift-ci please test and merge

**Pull request merged**